cmd/prestop/main.go

```diff
@@ -0,0 +1,231 @@
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return h, err
+	}
+	err = json.Unmarshal(data, &h)
+	return h, err
+}
+
+func ensureEnvironmentVariables(requiredEnvVars ...string) error {
+	var missingEnvVars []string
+	for _, envVar := range requiredEnvVars {
+		if val := os.Getenv(envVar); val == "" {
+			missingEnvVars = append(missingEnvVars, envVar)
+		}
+	}
+	if len(missingEnvVars) > 0 {
+		return fmt.Errorf("missing envars: %s", strings.Join(missingEnvVars, ","))
+	}
+	return nil
+}
+
+func main() {
+	fmt.Println("Calling pre-stop hook!")
+	cfg := zap.NewDevelopmentConfig()
+	if err := ensureEnvironmentVariables(logFilePathEnv, agentStatusFilePathEnv); err != nil {
+		zap.S().Fatal("Not all required environment variables are present: %s", err)
+		os.Exit(1)
+	}
+	cfg.OutputPaths = []string{
+		os.Getenv(logFilePathEnv),
+	}
+	log, err := cfg.Build()
+	if err != nil {
+		zap.S().Errorf("Error building logger config: %s", err)
+		os.Exit(1)
+	}
+	logger = log.Sugar()
+	health, err := getAgentHealthStatus()
+	if err != nil {
+		logger.Errorf("Error getting the agent health file: %s", err)
+	}
+
+	prettyPrint(health)
+	shouldDelete, err := shouldDeletePod(health)
+	fmt.Printf("shouldDeletePod=%t\n", shouldDelete)
+	logger.Debugf("shouldDeletePod=%t", shouldDelete)
+	if err != nil {
+		logger.Errorf("Error in shouldDeletePod: %s", err)
+	}
+
+	if shouldDelete {
+		if err := deletePod(); err != nil {
+			// We should not raise an error if the Pod could not be deleted. It can have even
+			// worst consequences: Pod being restarted with the same version, and the agent
+			// killing it immediately after.
+			logger.Errorf("Could not manually trigger restart of this Pod because of: %s", err)
+			logger.Errorf("Make sure the Pod is restarted in order for the upgrade process to continue")
+		}
+
+		// If the Pod needs to be killed, we'll wait until the Pod
+		// is killed by Kubernetes, bringing the new container image
+		// into play.
+		var quit = make(chan struct{})
+		logger.Info("A Pod killed itself, waiting...")
+		<-quit
+	}
+}
+
+// isWaitingToBeDeleted determines if the agent is currently waiting
+// on the mongod pod to be restarted. In order to do this, we need to check the agent
+// status file and determine if the mongod has been stopped and if we are in the process
+// of a version change.
+func isWaitingToBeDeleted(healthStatus agenthealth.MmsDirectorStatus) bool {
+	if len(healthStatus.Plans) == 0 {
+		return false
+	}
+	lastPlan := healthStatus.Plans[len(healthStatus.Plans)-1]
+	for _, m := range lastPlan.Moves {
+
+		// The next conditions are based on observations on the outcome
+		// of the agent after they have stopped the mongo server.
+
+		if m.Move == "WaitFeatureCompatibilityVersionCorrect" {
+			// First condition observed. This is the Plan reported by the
+			// agent on the first MongoD stopped.
+			for _, s := range m.Steps {
+				if s.Step == "WaitFeatureCompatibilityVersionCorrect" &&
```

Comment:
I've seen this stage coming up, why should we delete the pod when seeing `WaitFeatureCompatibilityVersionCorrect`. Does this require a restart?

Also maybe instead of 2 ifs, we could switch on `m.Move`

Reply:
this is only seen when the first mongod is stopped, my theory is that the the agent already moved to the next plan, which is "start" the mongodb now that it has been stopped.

After observing the agent's output and plan, this was consistent. Evey time the mongod was restarted, then the first one to be upgraded had this plan, while the following ones, had the other case in question: "ChangeVersion"